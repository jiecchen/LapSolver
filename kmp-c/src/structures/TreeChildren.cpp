#include "TreeChildren.h"

TreeChildren::TreeChildren(int nv, const int* parent) {
    // offset = cumulative sums of child counts
    offset = new int[nv+1];
    offset[0:nv+1] = 0;

    for (int i = 0; i < nv; i++) {
        if(parent[i] != i) offset[parent[i]+1]++;
    }
    for (int i = 1; i <= nv; i++) {
        offset[i] += offset[i-1];
    }

    // fill child array with these offsets
    child = new int[nv];
    int* childIndex = new int[nv];
    childIndex[0:nv] = 0;

    for (int i = 0; i < nv; i++) {
        int p = parent[i];
        if (p != i) {
            child[offset[p] + childIndex[p]++] = i;
        }
    }

    delete[] childIndex;
}

TreeChildren::~TreeChildren() {
    delete[] child;
    delete[] offset;
}