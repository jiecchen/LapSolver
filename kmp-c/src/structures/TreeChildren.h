// TreeChildren: a structure to compute and store child arrays from a parent array.
// child[offset[v], offset[v]+1, ..., offset[v+1]-1] are the children of vertex v.

#pragma once

struct TreeChildren {
    TreeChildren(int nv, const int* parent);
    ~TreeChildren();

    int* child;
    int* offset;
};
