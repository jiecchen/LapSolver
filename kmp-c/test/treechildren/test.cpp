// Test for TreeChildren

#include <cstdlib>
#include <iostream>
#include <string>
#include <structures/TreeChildren.h>

using namespace std;

int main(int argc, char const *argv[])
{
    // read a parent array from stdin
    const int MAX_N = 1000;
    int nv, parent[MAX_N], root;
    cin >> nv;
    for (int i = 0; i <= nv; i++) {
        cin >> parent[i];
        if (parent[i] == i) root = i;
    }

    printf("Root: %d\n", root);

    // convert to child array
    TreeChildren tChildren(nv, parent);

    for (int i = 0; i < nv; i++) {
        printf("%d:", i);
        for (int j = tChildren.offset[i]; j < tChildren.offset[i+1]; j++) {
            printf(" %d", tChildren.child[j]);
        }
        printf("\n");
    }

    return 0;
}
