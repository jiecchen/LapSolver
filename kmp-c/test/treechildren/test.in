15
2
2
8
8
8
9
10
10
11
11
12
14
14
14
14
