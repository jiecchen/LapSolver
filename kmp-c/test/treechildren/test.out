Root: 14
0:
1:
2: 0 1
3:
4:
5:
6:
7:
8: 2 3 4
9: 5
10: 6 7
11: 8 9
12: 10
13:
14: 11 12 13
